### PR TITLE
core(tbt): Clean up total blocking time code

### PR DESCRIPTION
**Pull request by deepanjanroy**
This PR cleans up Total Blocking Time code in lighthouse. It
- changes the name from CLQD -> TBT as decided by the [bikeshedding doc](https://docs.google.com/document/d/1SaYGOCy4_wEEoa0N_gNb1PyOxnZgSvuu_r52WhRYFKQ/edit#).
- Changes the definition of blocking time to be anything after first 50ms of a task, as opposed to anything before last 50ms. People [found](https://docs.google.com/a/google.com/document/d/1atjVa5eOuh1pGd0d5j3axDvj0jewN3FZZFESonnwu8A/edit?disco=AAAADQT1Gm8) the previous definition confusing, and this detail would change the value of the metric only by a maximum of 50ms, so the simplicity is worth it. 
- Has properly i18n-ed UI strings.

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -4280,6 +4280,11 @@ is-windows@^1.0.2:
   resolved "https://registry.yarnpkg.com/is-windows/-/is-windows-1.0.2.tgz#d1850eb9791ecd18e6182ce12a30f396634bb19d"
   integrity sha512-eXK1UInq2bPmjyX6e3VHIzMLobc4J94i4AWn+Hpq3OU5KkrRC96OAcR3PRJ/pGu6m8TRnBHP9dkXQVsT/COVIA==
 
+is-wsl@^1.1.0:
+  version "1.1.0"
+  resolved "https://registry.yarnpkg.com/is-wsl/-/is-wsl-1.1.0.tgz#1f16e4aa22b04d1336b66188a66af3c600c3a66d"
+  integrity sha1-HxbkqiKwTRM2tmGIpmrzxgDDpm0=
+
```
Comment (deepanjanroy) [inline yarn.lock]: I don't know what this is. Should I do anything about this?

File: lighthouse-core/lib/i18n/locales/en-XL.json [hunk 2]
```diff
@@ -1253,9 +1259,15 @@
   "lighthouse-core/lib/lh-error.js | dnsFailure": {
     "message": "D̂ŃŜ śêŕv̂ér̂ś ĉóûĺd̂ ńôt́ r̂éŝól̂v́ê t́ĥé p̂ŕôv́îd́êd́ d̂óm̂áîń."
   },
+  "lighthouse-core/lib/lh-error.js | erroredRequiredArtifact": {
+    "message": "R̂éq̂úîŕêd́ {artifactName} ĝát̂h́êŕêŕ êńĉóûńt̂ér̂éd̂ án̂ ér̂ŕôŕ: {errorMessage}"
+  },
```
Comment (deepanjanroy) [inline lighthouse-core/lib/i18n/locales/en-XL.json]: Not sure where this (and the blob below) came from.

Comment (brendankenny) [reply lighthouse-core/lib/i18n/locales/en-XL.json]: > Not sure where this (and the blob below) came from.

master was out of date. #9404 fixed this so it should disappear if you rebase or merge

File: yarn.lock (hunk 1 above)
Comment (brendankenny) [reply yarn.lock]: > I don't know what this is. Should I do anything about this?

oh, actually same deal. Fixed in #9404 :)

**Comment by brendankenny** [comment]
I'd like to get this into 5.2 so the next http archive run (starting August 1st) will have the up-to-date version of TBT

File: lighthouse-core/audits/metrics/total-blocking-time.js [hunk 3]
```diff
@@ -6,24 +6,28 @@
...
+const i18n = require('../../lib/i18n/i18n.js');
 
-// TODO(deepanjanroy): i18n strings once metric is final.
-const UIStringsNotExported = {
-  title: 'Cumulative Long Queuing Delay',
-  description: '[Experimental metric] Total time period between FCP and Time to Interactive ' +
-      'during which queuing time for any input event would be higher than 50ms.',
+const UIStrings = {
+  /** The name of the metric that calculates the total duration of blocking time for a page. Blocking times are time periods when the page would be slow at responding to user input (clicks, taps, and keypresses will feel janky). Shown to users as the label for the numeric metric value. Ideally fits within a ~40 character limit. */
+  title: 'Total Blocking Time',
```
Comment (patrickhulce) [inline lighthouse-core/audits/metrics/total-blocking-time.js]: @exterkamp did we decide we aren't translating metrics or just abbreviations? what's the deal now?

File: lighthouse-core/audits/metrics/total-blocking-time.js [hunk 4]
```diff
@@ -6,24 +6,28 @@
...
+  /** Description of the Total Blocking Time (TBT) metric, which calculates the total duration of blocking time for a page. Blocking times are times when if a page received any input (clicks, taps, and keypresses), the page would be slow at responding. This is displayed within a tooltip when the user hovers on the metric name to see more. No character length limits.*/
+  description: 'Sum of all time periods between FCP and Time to Interactive, ' +
+      'when task length exceeded by more than 50ms, expressed in milliseconds.',
```
Comment (patrickhulce) [inline lighthouse-core/audits/metrics/total-blocking-time.js]:
```suggestion
      'when task length exceeded 50ms, expressed in milliseconds.',
```

maybe?

Comment (patrickhulce) [reply lighthouse-core/audits/metrics/total-blocking-time.js]: or perhaps?

`Sum of all task execution time between FCP and TTI that occurred after the first 50ms of the task.`

forgive me if this was all coming from an official definition somewhere :)

File: lighthouse-core/computed/metrics/total-blocking-time.js [hunk 5]
```diff
@@ -39,30 +38,29 @@ class CumulativeLongQueuingDelay extends ComputedMetric {
...
       //   [              250ms Task                   ]
-      //   |  Long Queuing Delay Region  |   Last 50ms |
-      //               200 ms
-      longQueuingDelayRegions.push({
-        start: event.start,
-        end: event.end - threshold,
+      //   | First 50ms |     Blocking Time Region     |
+      //                            200 ms
+      blockingRegions.push({
+        start: event.start + threshold,
```
Comment (patrickhulce) [inline lighthouse-core/computed/metrics/total-blocking-time.js]: hey look at that it eventually circled around to make my old comment somewhat relevant! 😆

File: lighthouse-core/computed/metrics/total-blocking-time.js [hunk 6]
```diff
@@ -39,30 +38,29 @@ class CumulativeLongQueuingDelay extends ComputedMetric {
...
+    const threshold = TotalBlockingTime.BLOCKING_TIME_THRESHOLD;
+    const blockingRegions = [];
     for (const event of topLevelEvents) {
-      // If the task is less than the delay threshold, it contains no Long Queuing Delay Region.
+      // If the task is less than the delay threshold, it contains no Blocking Region.
       if (event.duration < threshold) continue;
-      // Otherwise, the duration of the task before the delay-threshold-sized interval at the end is
-      // considered Long Queuing Delay Region. Example assuming the threshold is 50ms:
+      // Otherwise, the duration of the task beyond blocking time threshold at the beginning is
+      // considered Blocking Region. Example assuming the threshold is 50ms:
```
Comment (patrickhulce) [inline lighthouse-core/computed/metrics/total-blocking-time.js]:
```suggestion
      // considered the Blocking Region. Example assuming the threshold is 50ms:
```

to match the below?

File: lighthouse-core/computed/metrics/total-blocking-time.js [hunk 7]
```diff
@@ -39,30 +38,29 @@ class CumulativeLongQueuingDelay extends ComputedMetric {
...
       //   [              250ms Task                   ]
-      //   |  Long Queuing Delay Region  |   Last 50ms |
-      //               200 ms
-      longQueuingDelayRegions.push({
-        start: event.start,
-        end: event.end - threshold,
+      //   | First 50ms |     Blocking Time Region     |
```
Comment (patrickhulce) [inline lighthouse-core/computed/metrics/total-blocking-time.js]: stick with just `Blocking Region`

File: lighthouse-core/computed/metrics/total-blocking-time.js [hunk 8]
```diff
@@ -39,30 +38,29 @@ class CumulativeLongQueuingDelay extends ComputedMetric {
...
         duration: event.duration - threshold,
       });
     }
 
-    let sumLongQueuingDelay = 0;
-    for (const region of longQueuingDelayRegions) {
-      // We only want to add up the Long Queuing Delay regions that fall between FCP and TTI.
+    let sumBlockingTime = 0;
+    for (const region of blockingRegions) {
+      // We only want to add up the Blocking Time Regions that fall between FCP and TTI.
```
Comment (patrickhulce) [inline lighthouse-core/computed/metrics/total-blocking-time.js]: not sure if it matters but is there significance to `Blocking Time Region` vs. `Blocking Region` or are they interchangeable?

File: lighthouse-core/test/computed/metrics/total-blocking-time-test.js [hunk 9]
```diff
@@ -72,27 +63,21 @@ Object {
...
+      ).toBe(150);
     });
 
-    it('clips queuing delay regions properly', () => {
-      const fcpTimeMs = 1050;
-      const interactiveTimeMs = 2050;
+    it('treats the end of task as blocking time', () => {
+      const fcpTimeMs = 100;
+      const interactiveTimeMs = 200;
+      // The last 50ms of this 100ms task is blocking time, and it happens after FCP, so it's
```
Comment (patrickhulce) [inline lighthouse-core/test/computed/metrics/total-blocking-time-test.js]: oh this is unexpected to me, there's no clipping involved at all? I would expect start of task to be clipped to `FCP` and so the "last 50" would still be free

File: lighthouse-core/test/computed/metrics/total-blocking-time-test.js [hunk 10]
```diff
@@ -72,27 +63,21 @@ Object {
...
+      // counted towards TBT.
+      const events = [{start: 50, end: 150, duration: 100}];
 
-      const events = [
-        {start: 1000, end: 1110, duration: 110}, // Contributes 10ms.
```
Comment (patrickhulce) [inline lighthouse-core/test/computed/metrics/total-blocking-time-test.js]: so under new model this is `// Contribtues 60ms`?

File: lighthouse-core/test/computed/metrics/total-blocking-time-test.js [hunk 11]
```diff
@@ -72,27 +63,21 @@ Object {
...
-        {start: 2000, end: 2200, duration: 200}, // Contributes 50ms.
```
Comment (patrickhulce) [inline lighthouse-core/test/computed/metrics/total-blocking-time-test.js]: and under the new model this is...still impossible? ;)

File: lighthouse-core/config/default-config.js [hunk 12]
```diff
@@ -369,7 +369,7 @@ const defaultConfig = {
         {id: 'first-cpu-idle', weight: 2, group: 'metrics'},
         {id: 'max-potential-fid', weight: 0, group: 'metrics'},
         {id: 'estimated-input-latency', weight: 0}, // intentionally left out of metrics so it won't be displayed
-        {id: 'cumulative-long-queuing-delay', weight: 0}, // intentionally left out of metrics so it won't be displayed
+        {id: 'total-blocking-time', weight: 0}, // intentionally left out of metrics so it won't be displayed
```
Comment (patrickhulce) [inline lighthouse-core/config/default-config.js]: pumped for when this comment isn't needed 😎

Comment (deepanjanroy) [reply lighthouse-core/config/default-config.js]: :boom:

File: lighthouse-core/computed/metrics/total-blocking-time.js [hunk 13]
```diff
@@ -39,50 +38,43 @@ class CumulativeLongQueuingDelay extends ComputedMetric {
...
+      if (event.start > interactiveTimeMs) continue;
 
-      // If a Long Queuing Delay Region spans the edges of our region of interest, we clip it to
-      // only include the part of the region that falls inside.
-      const clippedStart = Math.max(region.start, fcpTimeInMs);
-      const clippedEnd = Math.min(region.end, interactiveTimeMs);
-      const queuingDelayAfterClipping = clippedEnd - clippedStart;
+      // We first perform the clipping, and then calculate Blocking Region. So if we have a 150ms
+      // task [0, 150] and FCP happens midway at 50ms, we first clip the task to [50, 150], and then
+      // calculate the Blocking Region to be [100, 150]. There rational here is that tasks before
```
Comment (brendankenny) [inline lighthouse-core/computed/metrics/total-blocking-time.js]:
```suggestion
      // calculate the Blocking Region to be [100, 150]. The rational here is that tasks before
```

File: lighthouse-core/computed/metrics/total-blocking-time.js [hunk 14]
```diff
@@ -39,50 +38,43 @@ class CumulativeLongQueuingDelay extends ComputedMetric {
...
+      // FCP is unimportant, so we care whether the main thread is busy more than 50ms at a time
```
Comment (brendankenny) [inline lighthouse-core/computed/metrics/total-blocking-time.js]:
```suggestion
      // FCP are unimportant, so we care whether the main thread is busy more than 50ms at a time
```

File: lighthouse-core/audits/metrics/total-blocking-time.js [hunk 15]
```diff
@@ -6,24 +6,28 @@
...
 
-// TODO(deepanjanroy): i18n strings once metric is final.
-const UIStringsNotExported = {
-  title: 'Cumulative Long Queuing Delay',
-  description: '[Experimental metric] Total time period between FCP and Time to Interactive ' +
-      'during which queuing time for any input event would be higher than 50ms.',
+const UIStrings = {
+  /** The name of the metric that calculates the total duration of blocking time for a page. Blocking times are time periods when the page would be slow at responding to user input (clicks, taps, and keypresses will feel janky). Shown to users as the label for the numeric metric value. Ideally fits within a ~40 character limit. */
+  title: 'Total Blocking Time',
+  /** Description of the Total Blocking Time (TBT) metric, which calculates the total duration of blocking time for a page. Blocking times are times when if a page received any input (clicks, taps, and keypresses), the page would be slow at responding. This is displayed within a tooltip when the user hovers on the metric name to see more. No character length limits.*/
```
Comment (brendankenny) [inline lighthouse-core/audits/metrics/total-blocking-time.js]: (if the above suggestion seems good, reuse here, or reuse the one here up there instead :)

```suggestion
  /** Description of the Total Blocking Time (TBT) metric, which calculates the total duration of blocking time for a web page. Blocking times are time periods when the page would be blocked (prevented) from responding to user input (clicks, taps, and keypresses will feel slow to respond). This is displayed within a tooltip when the user hovers on the metric name to see more. No character length limits.*/
```

File: lighthouse-core/audits/metrics/total-blocking-time.js (hunk 3 above)
Comment (patrickhulce) [reply lighthouse-core/audits/metrics/total-blocking-time.js]: friendly ping on this one, from @brendankenny's request for better translator explanation I'm guessing we *do* want them translated afterall, so maybe it was just abbreviations I'm remembering?

File: lighthouse-core/test/audits/__snapshots__/metrics-test.js.snap [hunk 16]
```diff
@@ -35,5 +34,6 @@ Object {
   "observedTraceEndTs": 225426711887,
   "speedIndex": 1657,
   "speedIndexTs": undefined,
+  "totalBlockingTime": 726,
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/__snapshots__/metrics-test.js.snap]: now that the clipping order has moved back to the original, the fact that this changes is unexpected to me.

could we perhaps add a new unit test case that would be different between the two versions to capture this more explicitly? 

from my reading it should only be different in cases that I thought to be impossible

File: lighthouse-core/test/computed/metrics/total-blocking-time-test.js (hunk 11 above)
Comment (patrickhulce) [reply lighthouse-core/test/computed/metrics/total-blocking-time-test.js]: it seems like this test still exists just with different numbers, is that intentional? (the second task spans TTI)

File: lighthouse-core/test/computed/metrics/total-blocking-time-test.js [hunk 17]
```diff
@@ -72,27 +63,28 @@ Object {
...
-      const interactiveTimeMs = 2050;
+    it('clips before finding blocking regions', () => {
+      const fcpTimeMs = 150;
+      const interactiveTimeMs = 300;
 
       const events = [
-        {start: 1000, end: 1110, duration: 110}, // Contributes 10ms.
-        {start: 2000, end: 2200, duration: 200}, // Contributes 50ms.
+        // The clipping is done first, so the task becomes [150, 200] after clipping and contributes
+        // 0ms of blocking time. This is in contrast to first calculating the blocking region ([100,
```
Comment (patrickhulce) [inline lighthouse-core/test/computed/metrics/total-blocking-time-test.js]: would you mind adding a few other edge cases?

The fact that I can't figure out why this new version is different from before is my primary motivation :) 

I'm thinking...

```
          FCP at 1000 ms
          |
        =====  Task >50ms ends before 1050 ms
        ======  Task >50ms ends at 1050 ms
        =======  Task >50ms starts before 1000 ms and ends after 1050 ms

File: lighthouse-core/audits/metrics/total-blocking-time.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-core/audits/metrics/total-blocking-time.js]: ha, I did not consult @exterkamp when suggesting the description change. I think we might be punting on metrics/abbreviations until after placeholders (which will itself be after this PR)

File: lighthouse-core/audits/metrics/total-blocking-time.js [hunk 18]
```diff
@@ -6,24 +6,28 @@
...
+const ComputedTBT = require('../../computed/metrics/total-blocking-time.js');
+const i18n = require('../../lib/i18n/i18n.js');
 
-// TODO(deepanjanroy): i18n strings once metric is final.
-const UIStringsNotExported = {
-  title: 'Cumulative Long Queuing Delay',
-  description: '[Experimental metric] Total time period between FCP and Time to Interactive ' +
-      'during which queuing time for any input event would be higher than 50ms.',
+const UIStrings = {
+  /** The name of the metric that calculates the total duration of blocking time for a page. Blocking times are time periods when the page would be slow at responding to user input (clicks, taps, and keypresses will feel janky). Shown to users as the label for the numeric metric value. Ideally fits within a ~40 character limit. */
```
[the remark this answers is not in the document]
Comment (deepanjanroy) [reply lighthouse-core/audits/metrics/total-blocking-time.js]: Thanks for improving the wording! Changed comment.

File: lighthouse-core/audits/metrics/total-blocking-time.js (hunk 3 above)
Comment (deepanjanroy) [reply lighthouse-core/audits/metrics/total-blocking-time.js]: Should I be removing i18n code and comments for the title then, or is it fine to leave it as it is?

File: lighthouse-core/audits/metrics/total-blocking-time.js (hunk 15 above)
Comment (deepanjanroy) [reply lighthouse-core/audits/metrics/total-blocking-time.js]: Accepted.

File: lighthouse-core/test/computed/metrics/total-blocking-time-test.js (hunk 11 above)
Comment (deepanjanroy) [reply lighthouse-core/test/computed/metrics/total-blocking-time-test.js]: It was not intentional, but I just realized that TTI can happen at FMP or DCL, which can happen in the middle of a task, so this case is not as impossible as I thought. Testing this case is worth it now.

File: lighthouse-core/test/audits/__snapshots__/metrics-test.js.snap (hunk 16 above)
Comment (deepanjanroy) [reply lighthouse-core/test/audits/__snapshots__/metrics-test.js.snap]: Hm very good point - thanks for flagging this. I looked into it, and looks like the impossible is happening - TTI is in the middle of a task for the lantern test case (and it's not at FMP or DCL). Right now I'm trying to understand why that is the case. 

Side note: It looks like lantern TTI does not consider DCL. I see this at line 70 of lantern-interactive.js. Is this expected?
```js
    const lastTaskAt = LanternInteractive.getLastLongTaskEndTime(simulationResult.nodeTimings);
    const minimumTime = extras.optimistic
      ? extras.fmpResult.optimisticEstimate.timeInMs
      : extras.fmpResult.pessimisticEstimate.timeInMs;
    return {
      timeInMs: Math.max(minimumTime, lastTaskAt),
      nodeTimings: simulationResult.nodeTimings,
    };
```

File: lighthouse-core/audits/metrics/total-blocking-time.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-core/audits/metrics/total-blocking-time.js]: > Should I be removing i18n code and comments for the title then, or is it fine to leave it as it is?

no, I think should leave as it is here and we'll go through them in the future when we deal with all of them at once (after #9114 some time)

File: lighthouse-core/test/audits/__snapshots__/metrics-test.js.snap (hunk 16 above)
Comment (patrickhulce) [reply lighthouse-core/test/audits/__snapshots__/metrics-test.js.snap]: Oh, right because we are using the TTI estimate `timeInMs` of a potentially different graph so it is possible.

Re: DCL, Yeah there's not a reliable way we have to predict when DCL will occur under a hypothetical connection type so it's not considered.

AI from my perspective: we add a test case *just* exercising the TTI clipping behavior with a comment about why this can happen

Comment (deepanjanroy) [reply lighthouse-core/test/audits/__snapshots__/metrics-test.js.snap]: Ah right. That sgtm: added test case just for TTI clipping.

File: lighthouse-core/test/computed/metrics/total-blocking-time-test.js (hunk 17 above)
Comment (deepanjanroy) [reply lighthouse-core/test/computed/metrics/total-blocking-time-test.js]: Added these cases.

**Comment by patrickhulce** [comment]
@brendankenny did you explicitly *not* want it to land for 5.2?

**Comment by brendankenny** [comment]
> @brendankenny did you explicitly not want it to land for 5.2?

It depends on how @paulirish wants to do the branch? Presumably it could branch from any commit...

**Comment by patrickhulce** [comment]
Yeah I got that part :) I'm asking why you wouldn't want it to be included in 5.2 so we don't have to rush out a 5.3 before August 1?

**Comment by brendankenny** [comment]
lol oh I see, yeah that makes sense. Let's land :)
